preset: Additional presets and revisions for 1.6.0.

sr55 opened the pull request:
Additional changes to these added in #4757.

### Changes since 1.5.x

#### General

- Add `Very Fast 2160p60 4K AV1`
  - *AV1 video (up to 2160p60) and AAC stereo audio, in an MP4 container.*
- Add `Very Fast 2160p60 4K HEVC`
  - *HEVC video (up to 2160p60) and AAC stereo audio, in an MP4 container.*
- Add `Fast 2160p60 4K AV1`
  - *AV1 video (up to 2160p60) and AAC stereo audio, in an MP4 container.*
- Add `Fast 2160p60 4K HEVC`
  - *HEVC video (up to 2160p60) and AAC stereo audio, in an MP4 container.*
- Add `HQ 2160p60 4K AV1 Surround`
  - *High quality AV1 video (up to 2160p60), AAC stereo audio, and Dolby Digital (AC-3) surround audio, in an MP4 container.*
- Add `HQ 2160p60 4K HEVC Surround`
  - *High quality HEVC video (up to 2160p60), AAC stereo audio, and Dolby Digital (AC-3) surround audio, in an MP4 container.*
- Add `Super HQ 2160p60 4K AV1 Surround`
  - *Super high quality AV1 video (up to 2160p60), AAC stereo audio, and Dolby Digital (AC-3) surround audio, in an MP4 container.*
- Add `Super HQ 2160p60 4K HEVC Surround`
  - *Super high quality HEVC video (up to 2160p60), AAC stereo audio, and Dolby Digital (AC-3) surround audio, in an MP4 container.*


#### Web

- Rename `Vimeo YouTube HQ` → `Creator` (all presets)
  - *High quality video for publishing via online services such as Vimeo and YouTube. H.264 video (up to `resolution`) and high bit rate AAC stereo audio in an MP4 container.*
- Remove `Vimeo YouTube 720p30`
- Rename `Gmail` → `Email 25 MB` (all presets)
  - *Up to `duration` minutes of video in 25 MB or less, for sharing via email services such as Gmail. H.264 video (up to `resolution`) and AAC stereo audio, in an MP4 container.*
- Rename `Discord Nitro Large 3-6 Minutes 1080p30` → `Social 100 MB 5 Minutes 1080p30`
  - *Up to 5 minutes of video in 100 MB or less, for sharing via online social communities such as Discord. H.264 video (up to 1080p30) and AAC stereo audio, in an MP4 container.*
  - Increase average bit rate from 2100 kb/s to 2400 kb/s, maximum bit rate from 21000 kb/s to 24000 kb/s
- Rename `Discord Nitro Medium 5-10 Minutes 720p30` → `Social 50 MB 5 Minutes 720p30`
  - *Up to 5 minutes of video in 50 MB or less, for sharing via online social communities such as Discord. H.264 video (up to 720p30) and AAC stereo audio, in an MP4 container.*
  - Can also be used to encode 10 minutes of 720p30 video in 100 MB or less. Science!
- Rename `Discord Nitro Small 10-20 Minutes 480p30` → `Social 50 MB 10 Minutes 480p30`
  - *Up to 10 minutes of video in 50 MB or less, for sharing via online social communities such as Discord. H.264 video (up to 480p30) and AAC stereo audio, in an MP4 container.*
- Rename `Discord Small 2 Minutes 360p30` → `Social 8 MB 3 Minutes 360p30`
  - *Up to 3 minutes of video in 8 MB or less, for sharing via online social communities such as Discord. H.264 video (up to 360p30) and AAC stereo audio, in an MP4 container.*
  - Change encoder speed preset from `medium` to `slow`
  - Reduce average bit rate from 450 kb/s to 270 kb/s, maximum bit rate from 4500 kb/s to 2700 kb/s
  - The idea here is to consolidate the previous Discord Small and Discord Tiny presets; bit rate and duration limit are roughly a mid point between the two
- Remove `Discord Tiny 5 Minutes 240p30`


#### Devices

- Revise description for `Amazon Fire 2160p60 4K HEVC Surround` to specify compatibility with 3rd gen devices and newer instead of 2nd gen, which according to the manufacturer only supports up to 2160p30
- Revise description for `Apple 2160p60 4K HEVC Surround` to specify iPhone 7 and later, instead of a much longer list

#### Other

- Update all software HEVC (x265) presets to use 10-bit encoder and custom additional encoder options string, matching the new 4K general presets

<hr>

<details>
<summary>Original description (outdated)</summary>

New Presets:
 - Very Fast 2160p60 4K
 - Fast 2160p60 4K
 -  HQ 2160p60 4K
 -  Super HQ 2160p60 4K
    
Notes:
- The existing default preset of Fast 1080p30 remains as default since 1080p is still the dominant use case and allows for users to have to not worry about compatibility as much.

- I've also changed the audio bitrate on the 1080p Fast (192kbit), HQ (256kbit) and Super HQ presets (256kbit).  The 4K presets use the same bitrates for audio.

- Decomb and Interlace detection is turned OFF for 4K presets. The performance hit isn't work it and seemingly little 4K content is actually interlaced. 


Other Thoughts:

- Contemplating whether we should set Deinterlace detection to "Fast" for 1080p Presets out the box. It's defiantly considerably faster however I've not yet tested if it's viable with interlaced content. Will do so although if anyone has any experience on this let me know.


Open to any other ideas for preset tweaks while I'm at it.
</details>

bradleysepos commented:
Initial thoughts:

My goal with the general presets has been maximum compatibility across players and devices, and a reasonably fast default preset since video encoding is slow by nature. So of course, my comments will be through this lens. 😸

Need to verify whether > 160 kb/s audio is widely compatible. While I suspect it's fine for anything supporting 4K, regardless of manufacturer, Apple still lists this as the max for various devices despite them supporting higher, and iTunes is a real stickler about not syncing such files sometimes.

I'm not a fan of changing this for the existing 1080p presets. That said, if we go this route, I suggest maintaining 160 kb/s for Very Fast and Fast, and going to 256 kb/s for HQ and Super HQ (though some might argue the stereo track is a fallback track and this only bloats the file size). Keeps it simple, avoids changing the default preset, and means down the road if/when we switch the default to 4K, we're only changing video stuff.

Given how much archival stuff people do in general, I suspect changes to existing behavior will cause some pushback. I don't even want to think about benchmark comparisons ("but it has the same name!").

Before we disable interlace detection and removal outside of the hardware category, I'd like to see us implement what we've discussed in the past (and again recently on IRC), some smart behavior to only run these on <= 1080p. Most recently, I suggested doing this as a set of tunes for the detection filter, such as Always Detect, HD/SD Only, and SD Only. Custom cutoff resolution could be specified manually via custom options, of course. HD/SD Only would be a great default for all existing and new HandBrake presets > 1080p.

bradleysepos commented:
Looking at this further, I realize the 4K presets are AVC, not HEVC. That's a sticking point for me, I think we should be using x265 here, with additional options per the other 4K presets (probably disable rect only for the faster two presets).

Check video frame rate limit, it's 30 on at least one of the new ones.

sr55 commented:
Yeh, Lets leave the existing presets alone as it'll catch folks off guard that don't read the change note anyway. All Reverted.

I was debating the use of x265.  There was 3 reasons I went against that idea.

1. The reset of the General presets are AVC and we'd probably have to add HEVC into the preset name to make that clear if we wanted to do that.
2. iTunes on Windows is a no-go for HEVC the last time I checked so keeping with the compatibility theme I figured I'd stick with AVC for these.
3. x265 is about 1/3rd the speed which didn't really meet the "very fast" speed name. 


Happy to hold off until we get the Smart Comb Detect presets in place and update this post.


I'll need to try find some more interlaced content to test with but I suspect "Fast" interlace detect might not be good enough to use in the presets. (I.e at the 1080p level and below.)

bradleysepos commented:
> Yeh, Lets leave the existing presets alone as it'll catch folks off guard that don't read the change note anyway. All Reverted.

👍 

> I was debating the use of x265. There was 3 reasons I went against that idea.
> 
>     1. The reset of the General presets are AVC and we'd probably have to add HEVC into the preset name to make that clear if we wanted to do that.

Yes, we'd need to add HEVC to the preset name per the other presets.

>     2. iTunes on Windows is a no-go for HEVC the last time I checked so keeping with the compatibility theme I figured I'd stick with AVC for these.

That's odd, I wasn't aware of that. I was under the impression 4K AVC had compatibility issues. Will have to check my notes and external references on #1582 and prior work (linked there).

>     3. x265 is about 1/3rd the speed which didn't really meet the "very fast" speed name.

Yeah, very fast is harder to achieve there. Last I looked, you hadn't varied the encoder preset from medium, perhaps a much faster encoder preset can be tuned to produce reasonable results.

As much as I'm a fan of x264 and use it for pretty much everything, without the super transform introduced with H.265, I think the file size difference could be a bigger problem in the long run than speed.

> Happy to hold off until we get the Smart Comb Detect presets in place and update this post.

👍 

> I'll need to try find some more interlaced content to test with but I suspect "Fast" interlace detect might not be good enough to use in the presets. (I.e at the 1080p level and below.)

It's one of those things that's pretty ugly if we get it wrong, so perhaps go with the current defaults + the smart resolution bypass when implemented, and we can compare benchmarks and quality.

Thanks for your work on this! Apologies if I've come off a little *too* interested. Presets were kind of my baby for many years. Glad to have some help, especially as my time and abilities are more limited these days. Go team. 😸

bradleysepos commented:
Re-targeting for 1.6.0.

bradleysepos commented:
Rebased on current master and made some minor changes based on discussion and experimentation.

I've also been doing some more x265 tests to find a potentially better speed/size balance there. I don't think most users are going to be happy x264 yielding 25 Mbit/s video for a high quality 4K preset (likely pushing 40 Mbit/s for Super HQ). With x265 and appropriate encoder preset/settings, we can achieve ~50% performance for a 35-50% file size reduction. Tough call but I think the file size is going to be a bigger concern, especially at the higher quality end.

From my testing it's looking like for Super HQ, using roughly the same settings we use for existing 4K HEVC presets with a lower CRF value is spot on. Same encoding speed as existing 4K HEVC presets so no surprises. Ramp up the CRF and faster encoder presets for HQ, Fast, and Very Fast, so it only gets better from a performance standpoint. This would mean we won't be introducing any presets slower than ones we already have, and at the same time deliver smaller file sizes across the board and keep the higher quality presets from blowing out entirely.

Will push some additions here for others to try out once I'm done testing/tweaking.

bradleysepos commented:
Pushed HEVC versions for testing. Feel free to build and give these a whirl. Feedback is appreciated.

Some comparison using `tearsofsteel_4k.mov` on my old 12c/24t Westmere (nice because x264 and x265 saturate similarly at these settings), numbers are approximate:

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 27 | 15 min / 7 min | +114% | 4.31 M / 7.13 M | -40% |
| Fast CRF 25 | 23 min / 14 min | +64% | 6.65 M / 12.80 M | -48% |
| HQ CRF 23 | 43.5 min / 27 min | +61% | 9.87 M / 20.6 M | -52% |
| Super HQ CRF 21 | 90 min / 57 min | +58% | 17.3 M / 29.6 M | -42% |

Basically a ~40-50% file size reduction in exchange for ~60% longer encode times, except Very Fast which doubles encode time (can't do much better easily). File sizes for the high quality presets are more reasonable.

Further detail:

- HEVC and AVC CRF values are 27-21 and 26-20, respectively, in +/- 2 increments
- x265 encoder presets are faster to compensate for longer encoding times, and use the same custom settings as our other HEVC presets (`strong-intra-smoothing=0:rect=0:aq-mode=1`); other combinations of custom encoder settings have not proven significantly better in any way to change this approach
- In practice this means Very Fast will use AQ Mode 1 instead of 0 which actually improves blocking/banding in flat areas for a few percent speed hit, and Super HQ will not use rectangular motion partition analysis (same as other HEVC presets) which can be a significant performance hit on the order of 25%
- Overall the quality appears to be the same or better compared to x264 in nearly every case, so we're trading off time and size which was my goal here.
- As expected, x265 is smoother in the higher frequencies than x264, rejecting the finest grain. At least it's fairly uniform. Grain tuning was not employed for these comparisons
- While I generally prefer x264 for 1080p and below, here it exhibits more artifacts and grain weirdness such as very smooth patches next to very detailed or artifact-laden patches in broad flat areas... I suspect a further bump in bit rate would solve these issues, but as we're already creating 20-30 Mbit/s encodes here, it seems better to use an encoder more efficient at 4K resolution

I need to test some lower resolution input (`ToS-4k-1920.mov` would be good) to ensure acceptable results. We can always bump up the CRF by -1 to help compensate if that presents a better trade-off.

Edit: I should note here per my previous comment, the Super HQ preset is essentially identical to the our other 4K HEVC presets (e.g. `Apple 2160p60 4K HEVC Surround` and `H.265 MKV 2160p60`), only slightly higher quality at CRF 21 instead of 22. So this actually offers some faster HEVC choices via HQ, Fast, and Very Fast.

Edited to fix maths (more than once).

<details>
<summary>Follow-up testing (discarded results, tap/click to expand):</summary>

Testing a couple additional options, increasing Very Fast from CRF 27 to 28 brings more file size savings for minimal cost, will peek the quality more another time. Also testing whether increasing Super HQ from CRF 21 to 22 (identical to our other 4K HEVC presets) yields appreciable savings in relation to quality. I'm tempted to leave this one a bit larger so it's more "super" but we'll see.

Test results:

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 28 (+1) | 15 min / 7 min | +114% | 3.72 M / 7.13M | -48% |
| Fast CRF 25 (same) | 23 min / 14 min | +64% | 6.65 M / 12.80 M | -48% |
| HQ CRF 23 (same) | 43.5 min / 27 min | +61% | 9.87 M / 20.6 M | -52% |
| Super HQ CRF 22 (+1) | 83.5 / 57 min | +65% | 13.8 M / 29.6 M | -53% |

Tough call. Very Fast is a bit smaller but without any time savings. Super HQ is considerably smaller for minor time savings, but the quality is slightly worse upon close inspection (still high quality, but this preset is intended to remove reasonable doubt). While this yields some consistency in the numbers in this comparison table, I do not believe the result for the end user based on the intent of each preset is better.

Edit: I also tried using the ultrafast encoder preset to make the Very Fast preset faster while keeping the same additional options string for uniformity. It speeds things up about 20% but I'm not digging the quality at CRF 28; it's blotchy in detailed areas and really inconsistent overall. No big surprises there since it adds a number of restrictions while completely avoiding useful base-level analyses like scenecut and quarter pixel motion estimation. It's a little bit better at CRF 27, but IMHO still not worth it even for the lesser of these presets. More evidence the original table at the top of this post is probably the sweet spot.

</details>

Nomis101 commented:
> Further detail:
> 
> * HEVC and AVC CRF values are 27-21 and 26-20, respectively, in +/- 2 increments
> * x265 encoder presets are faster to compensate for longer encoding times, and use the same custom settings as our other HEVC presets (`strong-intra-smoothing=0:rect=0:aq-mode=1`); other combinations of custom encoder settings have not proven significantly better in any way to change this approach

I did nearly hundreds of tests with different custom encoder settings for x265 for material in 4k, HD and SD. And I can confirm your findings, that our current ones are very good. Aside from `rskip=2`, which improved the encoding time for 4k material significantly without visible loss in quality (the file size gets a bit bigger). And `hme` (hex,umh,star), which increased visible quality a little bit, but at the expense of speed.

bradleysepos commented:
> 2. iTunes on Windows is a no-go for HEVC the last time I checked so keeping with the compatibility theme I figured I'd stick with AVC for these.

Thinking about this point again. It would be great to confirm given iTunes on Windows is a reasonably a big target. Have we ever had complaints about it in relation to the Apple HEVC preset?

bradleysepos commented:
More comparison, this time using the 1080p source `ToS-4k-1920.mov` to ensure 4K HEVC presets aren't going to negatively affect lower resolution input. Again testing on my old 12c/24t Westmere (nice because x264 and x265 saturate similarly at these settings), numbers are approximate:

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 27 | 4.5 min / 2 min | +125% | 1.34 M / 1.83 M | -27% |
| Fast CRF 25 | 7.5 min / 3.5 min | +114% | 1.50 M / 2.92 M | -49% |
| HQ CRF 23 | 11 min / 6 min | +83% | 2.22 M / 3.66 M | -39% |
| Super HQ CRF 21 | 21.5 min / 12 min | +79% | 3.40 M / 4.65 M | -27% |

And another comparison against the existing 1080p30 presets for good measure:

| HB Preset | Encode Duration HEVC/AVC 1080p | Encode Duration vs. AVC 1080p | Video Bit Rate HEVC/AVC 1080p | Video Bit Rate vs. AVC 1080p |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 27 | 4.5 min / 2 min | +125% | 1.34 M / 2.44 M | -45% |
| Fast CRF 25 | 7.5 min / 3.5 min | +114% | 1.50 M / 3.82 M | -60% |
| HQ CRF 23 | 11 min / 6.5 min | +69% | 2.22 M / 4.84 M | -54% |
| Super HQ CRF 21 | 21.5 min / 12.5 min | +72% | 3.40 M / 6.05 M | -44% |

Generally speaking, I find x264 does a better job at quality here, in this area of resolution where it shines. x265 clearly allocates bits differently, perhaps more consistently across the frame, which means it looks a bit worse in still comparisons but is about the same in motion (though, not usually better). I'm not sure we can make this much better without throwing away file size reduction or drastically increasing the encode duration. x265 is and has always been optimized for > 1080p content (and ultra-low bit rate encodes, which is not the goal here).

bradleysepos commented:
I tried `rskip=2` with some faster encoder presets and found it to be a little bit faster. Though I did find in some cases there were pronounced artifacts in flat areas, so perhaps the quality is partially dependent on the other settings involved. Might work best with slower encoder presets.

Initial testing with `selective-sao=2` (read: disable sao for b-frames) yields a ~5% speedup without other bad effects. That's a newer feature I think I looked at awhile back but I'm not sure whether I did any extensive testing. If it's beneficial across the board, we could include it in our other presets. Curious why x265 didn't add it to their presets after [revealing good results at SMPTE 2019](https://thebroadcastknowledge.com/2020/01/30/video-speed-distortion-optimization-tradeoffs-in-open-source-hevc-encoding/). I haven't seen much benefit to the other mentioned setting, `dynamic-rd`, in limited testing. Both are mostly meant to speed up the slowest encoder presets, but at least there seems to be some universal benefit to selective sao. Will queue up some more encodes for analysis.

Update: `selective-sao=2` did not change encoding time at all for Fast through Super HQ at 4K resolution, and slowed down Very Fast by about 6.5%. Presumably the latter is because the superfast encoder preset disables sao, so in that case I was enabling some analyses instead of restricting them. *sigh* Another dead end.

bradleysepos commented:
I may have cracked the code: `strong-intra-smoothing=0:rect=0:aq-mode=1:rd=4:psy-rd=0.75:psy-rdoq=4.0:rdoq-level=1:rskip=2` and +1 CRF. Tested with small sections of ToS 4K and BB Sunflower 4K (seems to be fine on animation). Have some full encodes queued now.

Breaking it down:

- `psy-rd` and `psy-rdoq` are psycho-visual optimizations that aren't enabled until `rd=3` or higher.
- The faster presets use `rd=2` so we need to increase this in order to use these psy opts. There's not much difference between using `rd=3` and `rd=4` with the faster encoder presets, so `rd=4` ensures we're not limiting the Super HQ preset which uses the `slow` encoder preset which uses `rd=4` by default.
- `rdoq-level` is also needed for `psy-rdoq` to work.
- With these settings enabled (and tuned for general use), speed goes down and bit rate goes up; we add `rskip=2` to compensate speed (hat tip @Nomis101) and increase CRF by 1-2 to compensate bit rate.
- The result is much better retention of higher frequency detail with dramatically reduced banding in flat areas and background. It looks more like x264 in its sweet spot. Main quality nit might be faint speckled artifacts in some flat areas, but this seems easily outweighed by the overall improvement.
- HQ and Super HQ presets may end up being quite a bit faster (19% and 41% with a short test clip containing mixed motion) for a small size increase (about 6%). At 4K that should put Super HQ in roughly the same speed class as the x264 version, with improved quality and file size. Fast will probably be a little bit slower and Very Fast will probably be considerably slower, more like Fast without these additional settings. `selective-sao=2` did not help in either case.

To test, select one of the 4K HEVC presets added in [bb9bd8a](https://github.com/HandBrake/HandBrake/pull/3919/commits/bb9bd8a188f40cd8954c431df232bcadaaf9fb21). Set the video encoder advanced options to `strong-intra-smoothing=0:rect=0:aq-mode=1:rd=4:psy-rd=0.75:psy-rdoq=4.0:rdoq-level=1:rskip=2` and increase CRF by +1.

For a relatively direct comparison, Fast @ CRF 27 (+2) with these settings seems to produce similar size at similar speed as Fast @ CRF 25 as committed. **Edit:** CRF 26.5 is closer.

Will post some numbers later (and perhaps some image comparison at some point) when my queue is empty.

Nomis101 commented:
Interesting. It looks different from the settings I've made up, but I'm excited to test it on my regular test samples. I have a few tricky test candidates where many settings I tried often gave me weird artifacts in the dark background or with fast movements.

bradleysepos commented:
New comparison updating [the original x265 vs x264 table](https://github.com/HandBrake/HandBrake/pull/3919#issuecomment-1107938141). Encoding `tearsofsteel_4k.mov` on the 12c/24t Westmere again, this time with psy opts enabled (`strong-intra-smoothing=0:rect=0:aq-mode=1:rd=4:psy-rd=0.75:psy-rdoq=4.0:rdoq-level=1:rskip=2`) and increasing CRF by +1 per [my previous post](https://github.com/HandBrake/HandBrake/pull/3919#issuecomment-1114191899):

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 28 | 19 min / 7 min | +171% | 4.33 M / 7.13 M | -39% |
| Fast CRF 26 | 27min / 14 min | +93% | 6.90 M / 12.80 M | -46% |
| HQ CRF 24 | 37.5 min / 27 min | +39% | 10.2 M / 20.6 M | -50% |
| Super HQ CRF 22 | 60 min / 57 min | +5% | 16.0 M / 29.6 M | -46% |

Detail is considerably better for Very Fast in most every aspect, and seemingly more consistent across the frame. This comes at a 26% performance penalty over the previous settings. File size is about the same. Even though Very Fast is intended to be optimized for speed, x265 already struggles badly here and it looks like we can't create as speedy of a preset as we'd like without hosing quality completely. As Scott noted on IRC, the matching x264 preset in this PR actually challenges x265's quality here, and more bit rate does not seem to help. So this seems like a good trade to remove some obvious artifacts and bring the baseline quality up. Perhaps it's better to think of the Very Fast HEVC preset as 30% faster than the Fast HEVC preset, rather than compare to the AVC preset.

Overall detail and quality are a little bit better on Fast, though a more subtle improvement. Seems to diminish blocking and banding artifacts while refining edge structure and minor details in the background for a little more uniformity. 17% slower. File size is about the same. Still reasonably fast, perhaps as much as twice as fast as the Apple 4K preset and it looks pretty good. The Fast presets have always tried to strike a balance of good quality in reasonable time, and this seems to fit for this encoder.

Quality is slightly better for HQ, mostly in texture acutance and background details that were previously unrefined. Super HQ can show this in some areas as well but it's mostly the same. The real news here is performance is greatly improved. HQ got almost 14% faster and Super HQ only took 60 minutes here instead of 90 minutes, a full 33% faster than the previous settings and only 5% slower than x264, while keeping video bit rate much lower.

~Here are some comparison images of AVC, HEVC, and HEVC with psy opts enabled.~ Removed (obsolete).

All in all, this is probably good enough, though I feel like Super HQ (perhaps HQ also) could be a little better. Perhaps I'll start another queue adjusting CRF by -1 again...

**Update:**

The faster presets seem dialed pretty well for CRF with these settings. The higher quality presets might benefit from lowering CRF by -1. Screenshots added to the comparison linked above (`tos4k-2-hq-hevc-2.png` and `tos4k-3-superhq-hevc-2.png`).

Detail for HQ at CRF 23 (-1) with these settings holds up better to pixel peeping, with a 40.5 min encode duration (8% slower) and 12.6 Mb/s video bit rate (23.5% larger, still 39% smaller than the AVC version). What I mostly like about this is it cements HQ as essentially unobjectionable.

Improvement for Super HQ at CRF 21 (-1) with these settings is more subtle but it's there, at 62 min encode duration (3% slower) and 20.0 Mb/s video bit rate (25% larger, still 32% smaller than the AVC version). Tears of Steel 4K is letterboxed; if it was not, this would translate to about 25 Mb/s video bit rate for full-frame 3840x2160 4K in HEVC vs 37.5 Mb/s for AVC. Seems fair enough. I'll push these changes to make it easier to test.

Updated table:

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 28 | 19 min / 7 min | +171% | 4.33 M / 7.13 M | -39% |
| Fast CRF 26 | 27min / 14 min | +93% | 6.90 M / 12.80 M | -46% |
| **HQ CRF 23** | **40.5 min** / 27 min | **+50%** | **12.6 M** / 20.6 M | **-39%** |
| **Super HQ CRF 21** | **63 min** / 57 min | **+11%** | **20.0 M** / 29.6 M | **-32%** |

Nomis101 commented:
> To test, select one of the 4K HEVC presets added in [bb9bd8a](https://github.com/HandBrake/HandBrake/pull/3919/commits/bb9bd8a188f40cd8954c431df232bcadaaf9fb21). Set the video encoder advanced options to `strong-intra-smoothing=0:rect=0:aq-mode=1:rd=4:psy-rd=0.75:psy-rdoq=4.0:rdoq-level=1:rskip=2` and increase CRF by +1.

I have tried this with some of my samples. I'm not quite finished. Overall its quite good. The file size is very small and the encoding quite speedy. But in some cases I see some colored artifacts and banding in dark areas. One sample is this. The top is encoded with your Super HQ 4k preset and your mentioned advanced encoder options. The second one is encoded with my 4k preset (1) and the last one is the original file (its a bit brighter  in color because the top ones are copied from Quicktime and the last one from VLC).
![Unbenannt-5](https://user-images.githubusercontent.com/13453351/166835434-ccbebfc2-3da9-4849-ae29-2ea69eec8451.jpg)

In the SHQ one you can see on the right side in the dark a banding issue, in the face there is a red colored artifact above the right eyebrow and to my eyes, the details in the skin are a bit softened. I need to say, that in the picture there is a dark color artifact on the left side on the skin, which is not there in the video (no idea where this does come from).
However, the SHQ encoding was 3x faster than the one with my preset and the file size is also a lot smaller (I optimized mine only for quality, not so much for speed or file size).
I've also noticed, it is not a good idea to use your 4K presets for anything other than 4k. 

There is also one thing I noticed with your 4k presets. If it is a movie with a 7.1 EAC3 audio stream, it always gets downmixed to AC3 5.1. Is this necessary? Wouldn't it be a problem to passthu the original (E)AC3 stream, if one is present?


(1): For my encoding I've used preset "slower", CRF=20
and `early-skip:rskip=2:rect=0:amp=0:sao=0:rd-refine:rc-lookahead=25:opt-cu-delta-qp:strong-intra-smoothing=0:psy-rdoq=0:hme:hme-search=hex,umh,star:b-adapt=1:aq-strength=0.9:qcomp=0.75:ipratio=1.2:pbratio=1.1:auto-aq:psy-rd=3`

bradleysepos commented:
Great feedback, thanks. That's certainly worse banding than I've observed in my limited testing. Interesting. I'll have to see if I can dig up some more difficult sources. I actually thought these settings decreased banding but perhaps that's not universal.

It would be good to know what the video bit rate is compared to your preset (which looks quite good, by the way). Super HQ HEVC here is around 5x slower than real time. We could go slower if it significantly helps quality, but I'm not sure we want to approach 15x slower overall. I haven't yet had the chance to test lower resolution material, but it's on my list. I'll give your settings some attention as well.

Regarding surround formats, we had a discussion about this around the time we created the other HEVC 4K presets like the Apple and Roku ones, and agreed to the current settings for maximum compatibility with existing home theater equipment. When newer playback devices arrived on the market supporting newer formats, the surround processors/amps they were connected to often did not (TOSLINK optical transport was still a thing for a number of folks, myself included). I'm certainly open to revisiting the idea as more time has passed.

Nomis101 commented:
Sorry, it was late yesterday and I hadn't noticed that your presets are 8-bit. My samples are HDR and of course you get banding when you convert 10-bit to 8-bit. I did the whole thing again with 10-bit and now the artefacts are almost gone. There are still a few small coloured artefacts in the face and the dark background looks a bit soft ironed. But the quality is pretty good in terms of file size and speed.
Is there a reason why the 4k presets are 8-bit? I don't know the whole BlueRay market, but all the 4k (UltraHD) BlueRays I own are actually in HDR10.

About the bit rates you asked:
Original: 46.0 Mb/s
Your SHQ (after changing to 10bit): 5114 kb/s
Mine: 20.2 Mb/s

bradleysepos commented:
> Is there a reason why the 4k presets are 8-bit?

The initial 4K presets we created many years ago are 8-bit because we had some weird issues at the time, both concerning inflated output size using x265, and Apple TV 4K devices showing a green/black screen in some instances (possibly an early firmware issue). Regarding the former, while it's true 10-bit encodes should be smaller due to reduced rounding error in the quantization process, it's a roll of the dice in practice as encoders often employ more tricks than basic compression.

I did not take HDR into account while working on these, but we should. There also might be an argument for not moving forward with 4K AVC presets, since HDR basically requires 10-bit, and 10-bit AVC has traditionally had poor device support.

I'll do some testing with the x265 10-bit encoder and see if anything needs adjusted. If all indications are positive, we should definitely move forward with 10-bit HEVC. Perhaps we should also update the existing 4K device presets in tandem.

bradleysepos commented:
I did a bit more testing on the 8-bit encodes. Based on visual inspection, VMAF, and some test encodes on another faster machine, it seems there is some benefit to slightly bumping up the quality for the faster presets, as I did with the higher quality presets in https://github.com/HandBrake/HandBrake/pull/3919#issuecomment-1114467103. This adds about 90 secs encoding time on the older machine.

Updated table:

| HB Preset | Encode Duration HEVC/AVC | Encode Duration vs. AVC | Video Bit Rate HEVC/AVC | Video Bit Rate vs. AVC |
| --------- | --------- | --------- | --------- | --------- |
| Very Fast CRF 27 | 20.5 min / 7 min | +193% | 5.08 M / 7.13 M | -29% |
| Fast CRF 25 | 28.5 min / 14 min | +104% | 8.25 M / 12.80 M | -36% |
| HQ CRF 23 | 40.5 min / 27 min | +50% | 12.6 M / 20.6 M | -39% |
| Super HQ CRF 21 | 63 min / 57 min | +11% | 20.0 M / 29.6 M | -32% |

bradleysepos commented:
Testing 10-bit HEVC again, I've found there to be negligible difference in output size compared to earlier versions of x265. To ensure HDR compatibility (read: no banding), I suggest we make the 4K HEVC presets 10-bit, and now might be a good time to update others such as the Apple 4K preset (along with the new encoder options string, perhaps).

I've also spent more time encoding a wider array of sources, checking VMAF, and comparing against other codecs. I bumped the HEVC quality yet again by -1 CRF, which makes quality generally more robust for animation and some edge cases where x265 struggles. So the proposed 4K HEVC presets are CRF 26/24/22/20, same as the proposed 4K AVC presets. The HQ presets will use more data, but it seems to be the only way and I'm tired of trying to work around x265's deficiencies. If people want even smaller files, they can adjust the quality setting as usual. The biggest negative here is x265's CRF 20 is up to 50% slower than CRF 21, though this is probably less of a problem on more modern hardware.

We can also add 10-bit SVT-AV1 presets as well, if there are no objections. I did a copious amount of testing and I believe CRF 35/30/25/20 with presets 9/8/7/6 to be an acceptable speed/quality tradeoff. I recognize presets 4-5 are generally considered good starting points, but in real-world testing, I've found they severely limit performance (especially parallel computation) for far less quality/size benefit than would be expected.

Testing again on the old 12c/24t Westmere Xeon:

| HB Preset                          | Quality   | Encode Duration | Video Bit Rate | Encode Duration vs. AVC | Video Bit Rate vs. AVC |
| ---------------------------------- | --------- | --------------- | -------------- | ----------------------- | ---------------------- |
| Very Fast 2160p60 4K AV1           | CRF 35    | 23.5 min        | 5.8 M          | +236%                   | -18%                   |
| Very Fast 2160p60 4K HEVC          | CRF 26    | 25  min         | 6.0 M          | +257%                   | -15%                   |
| Very Fast 2160p60 4K               | CRF 26    | 7 min           | 7.1 M          |                         |                        |
| Fast 2160p60 4K AV1                | CRF 30    | 41.5 min        | 8.6 M          | +196%                   | -33%                   |
| Fast 2160p60 4K HEVC               | CRF 24    | 37 min          | 10.5 M         | +164%                   | -18%                   |
| Fast 2160p60 4K                    | CRF 24    | 14 min          | 12.8 M         |                         |                        |
| HQ 2160p60 4K AV1 Surround         | CRF 25    | 76 min          | 13.0 M         | +181%                   | -37%                   |
| HQ 2160p60 4K HEVC Surround        | CRF 22    | 53 min          | 16.5 M         | +96%                    | -20%                   |
| HQ 2160p60 4K Surround             | CRF 22    | 27 min          | 20.6 M         |                         |                        |
| Super HQ 2160p60 4K AV1 Surround   | CRF 20    | 184 min         | 19.1 M         | +223%                   | -35%                   |
| Super HQ 2160p60 4K HEVC Surround  | CRF 20    | 94 min          | 26.0 M         | +65%                    | -12%                   |
| Super HQ 2160p60 4K Surround       | CRF 20    | 57 min          | 29.6 M         |                         |                        |

10-bit AVC is simply not widely supported. Even though HDR is separate from 4K, if we move forward with these presets using 8-bit x264, we should expect a higher incidence of complaint about banding (or even general 4K AVC compatibility). I am okay with ditching 4K AVC general presets altogether and moving forward with only 4K HEVC/AV1 general presets, but this is of course open for discussion.

bradleysepos commented:
Rebased on master, updated 4K HEVC general presets to CRF 26/24/22/20 and 10-bit, added 4K AV1 10-bit general presets.

FlaTechNole21 commented:
I see one issue with the 10-bit x265 presets:
aq-mode should be 2, not 1. That mode is optimized for 10-bit and results in a lower bitrate with no noticable quality loss when using the same options otherwise. For 8-bit x265, 1 is good as the default.

bradleysepos commented:
> I see one issue with the 10-bit x265 presets:
> aq-mode should be 2, not 1. That mode is optimized for 10-bit and results in a lower bitrate with no noticable quality loss when using the same options otherwise. For 8-bit x265, 1 is good as the default.

Thanks, this is something I can look at further. The backstory is we had a lot of complaints when the x265 default was changed from 1 to 2, and I was able to verify 20-30% increased file sizes in some cases for very little (if any) quality gains, so we decided to continue onward with aq-mode 1. I did a few tests with aq-mode 2 and 10-bit here but don't recall what my observations were; will give it another look.

bradleysepos committed:
preset: Add 4K AV1 and HEVC general presets.

Co-authored-by: sr55 <sr55.code@outlook.com>

bradleysepos committed:
preset: Add some non-critical keys that were missing for some presets.

bradleysepos committed:
preset: Make existing HEVC presets use 10-bit encoder.

Update additional encoder options to match new 4K general presets.

bradleysepos committed:
preset: Correct description for Fire 4K preset.

Specify compatibility with 3rd gen devices and newer.

According to the manufacturer's documentation, 2nd gen devices only
support up to 30 fps; the preset allows up to 60 fps.

bradleysepos committed:
preset: Revise Web presets.

Changed

- Rename Vimeo YouTube HQ presets Creator
- Rename Gmail presets Email 25 MB
- Rename Discord presets Social 100 MB and Social 50 MB
  - Make some adjustments to durations and bit rates

Removed

- Remove Vimeo YouTube 720p30 preset
- Remove Discord Tiny 5 Minutes 240p30 preset

bradleysepos committed:
preset: Bump internal presets major revision number.

bradleysepos commented:
> > I see one issue with the 10-bit x265 presets:
> > aq-mode should be 2, not 1. That mode is optimized for 10-bit and results in a lower bitrate with no noticable quality loss when using the same options otherwise. For 8-bit x265, 1 is good as the default.
> 
> Thanks, this is something I can look at further. The backstory is we had a lot of complaints when the x265 default was changed from 1 to 2, and I was able to verify 20-30% increased file sizes in some cases for very little (if any) quality gains, so we decided to continue onward with aq-mode 1. I did a few tests with aq-mode 2 and 10-bit here but don't recall what my observations were; will give it another look.

I gave this another look and concluded it does not make much if any difference, specifically given the rest of the parameters we are passing here via the custom encoder options string.

bradleysepos commented:
I've pushed a consolidated set of commits adding 4K AV1 and HEVC general presets only, as we decided AVC would be too problematic given limited support for 10-bit depth required for HDR pass through. Along with some other minor changes this is essentially the revision set slated for HandBrake 1.6.0, so I will update the title and description accordingly.

galad87 commented:
The PictureAutoCrop key is not used anymore, and can be safely deleted.

bradleysepos commented:
> The PictureAutoCrop key is not used anymore, and can be safely deleted.

I was under the impression it was still needed for some compatibility reason. Not true?

galad87 commented:
I removed the last references in 60394848b852745f3229a0a09e7556f5070ceb67 , and since 9e2aee31d56794d2e16d8096d8e24278d19d8d4f it imports the value in the old presets to the new key. There shouldn't be any code left to use the PictureAutoCrop key.

bradleysepos commented:
> I removed the last references in [6039484](https://github.com/HandBrake/HandBrake/commit/60394848b852745f3229a0a09e7556f5070ceb67) , and since [9e2aee3](https://github.com/HandBrake/HandBrake/commit/9e2aee31d56794d2e16d8096d8e24278d19d8d4f) it imports the value in the old presets to the new key. There shouldn't be any code left to use the PictureAutoCrop key.

Great, thanks for clarifying. I'll remove it.

bradleysepos committed:
preset: Remove deprecated PictureAutoCrop key from presets.

bradleysepos commented:
> > I removed the last references in [6039484](https://github.com/HandBrake/HandBrake/commit/60394848b852745f3229a0a09e7556f5070ceb67) , and since [9e2aee3](https://github.com/HandBrake/HandBrake/commit/9e2aee31d56794d2e16d8096d8e24278d19d8d4f) it imports the value in the old presets to the new key. There shouldn't be any code left to use the PictureAutoCrop key.
> 
> Great, thanks for clarifying. I'll remove it.

Done.

I am working on some additional changes to the Web presets (naming, mostly) as discussed privately, and perhaps some description updates. Then this will be ready.

bradleysepos commented:
> I am working on some additional changes to the Web presets (naming, mostly) as discussed privately, and perhaps some description updates. Then this will be ready.

Done. Rebased and merged.
